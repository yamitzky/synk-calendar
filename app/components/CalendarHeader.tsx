import { Button, ButtonGroup, Select, SelectItem } from '@nextui-org/react'
import { subDays } from 'date-fns'
import { twMerge } from 'tailwind-merge'
import type { CalendarViewType } from '~/components/viewType'

type Props = {
  todayLabel?: string
  start?: string
  end?: string // exclusive
  onToday?: () => void
  onNext?: () => void
  onPrev?: () => void
  initialView?: CalendarViewType
  onChangeView?: (viewType: CalendarViewType) => void
  className?: string
}

export const CalendarHeader = ({
  start,
  end,
  onToday,
  onNext,
  onPrev,
  onChangeView,
  initialView,
  todayLabel,
  className,
}: Props) => {
  return (
    <div
      className={twMerge(
        'grid gap-2 grid-cols-[repeat(2,max-content)] sm:grid-cols-[repeat(3,max-content)] justify-between items-center',
        className,
      )}
    >
      <ButtonGroup>
        <Button onClick={onToday} className="px-4">
          {todayLabel ?? getToday()}
        </Button>
        <Button isIconOnly onClick={onPrev}>
          ◀
        </Button>
        <Button isIconOnly onClick={onNext}>
          ▶
        </Button>
      </ButtonGroup>
      <h1 className="text-lg text-center col-span-2 sm:col-span-1">{formatRange(start, end)}</h1>
      <div className="w-40 row-start-1 col-start-2 sm:col-start-3">
        <Select
          defaultSelectedKeys={initialView ? [initialView] : undefined}
          disallowEmptySelection
          aria-label="view type"
          onChange={(e) => onChangeView?.(e.target.value as CalendarViewType)}
        >
          <SelectItem key="dayGridMonth">月</SelectItem>
          <SelectItem key="timeGridWeek">週</SelectItem>
          <SelectItem key="timeGridFourDay">4日</SelectItem>
          <SelectItem key="timeGridDay">日</SelectItem>
        </Select>
      </div>
    </div>
  )
}

function formatRange(start?: string, end?: string): string {
  const startDate = start ? new Date(start) : new Date()
  const endDate = end ? subDays(new Date(end), 1) : new Date()

  const format = new Intl.DateTimeFormat('ja', {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
  })
  return format.formatRange(startDate, endDate)
}

function getToday() {
  const format = new Intl.DateTimeFormat('ja', {
    month: 'short',
    day: 'numeric',
  })
  return format.format(new Date())
}
