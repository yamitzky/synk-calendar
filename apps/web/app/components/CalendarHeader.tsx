import { Button, ButtonGroup, Input, Select, SelectItem } from '@nextui-org/react'
import type { User } from '@synk-cal/core'
import { subDays } from 'date-fns'
import { useState } from 'react'
import { twMerge } from 'tailwind-merge'
import { UserInfo } from '~/components/UserInfo'
import type { CalendarViewType } from '~/components/viewType'
import useLocale from '~/hooks/useLocale'

type Props = {
  todayLabel?: string
  user?: User
  start?: string
  end?: string // exclusive
  onToday?: () => void
  onNext?: () => void
  onPrev?: () => void
  onSearch?: (query: string) => void
  initialView?: CalendarViewType
  onChangeView?: (viewType: CalendarViewType) => void
  className?: string
}

export const CalendarHeader = ({
  user,
  start,
  end,
  onToday,
  onNext,
  onPrev,
  onChangeView,
  onSearch,
  initialView,
  todayLabel,
  className,
}: Props) => {
  const locale = useLocale()
  const langCode = locale?.split('-')?.[0]
  const [isEditing, setIsEditing] = useState(false)
  const [query, setQuery] = useState('')
  return (
    <div
      className={twMerge(
        'grid gap-2 grid-cols-[repeat(2,max-content)] sm:grid-cols-[repeat(3,max-content)] justify-between items-center',
        className,
      )}
    >
      <ButtonGroup>
        <Button onClick={onToday} className="px-4" aria-label="Today">
          {todayLabel ?? getToday(locale)}
        </Button>
        <Button isIconOnly onClick={onPrev} aria-label="Previous">
          ‚óÄ
        </Button>
        <Button isIconOnly onClick={onNext} aria-label="Next">
          ‚ñ∂
        </Button>
      </ButtonGroup>
      <h1 className="text-lg text-center col-span-2 sm:col-span-1">{formatRange(locale, start, end)}</h1>
      <div className="row-start-1 col-start-2 sm:col-start-3 flex space-x-1">
        <Select
          defaultSelectedKeys={initialView ? [initialView] : undefined}
          disallowEmptySelection
          aria-label="view type"
          onChange={(e) => onChangeView?.(e.target.value as CalendarViewType)}
          className="w-40"
        >
          <SelectItem key="dayGridMonth" data-testid="dayGridMonth">
            {langCode === 'ja' ? 'Êúà' : 'Month'}
          </SelectItem>
          <SelectItem key="timeGridWeek" data-testid="timeGridWeek">
            {langCode === 'ja' ? 'ÈÄ±' : 'Week'}
          </SelectItem>
          <SelectItem key="timeGridFourDay" data-testid="timeGridFourDay">
            {langCode === 'ja' ? '4Êó•' : '4 Days'}
          </SelectItem>
          <SelectItem key="timeGridDay" data-testid="timeGridDay">
            {langCode === 'ja' ? 'Êó•' : 'Day'}
          </SelectItem>
        </Select>

        {!onSearch ? null : isEditing ? (
          <Input
            isClearable
            startContent="üîç"
            className="w-40"
            autoFocus
            placeholder="Search..."
            value={query}
            onChange={(e) => setQuery(e.target.value)}
            onKeyDown={(e) => {
              if (e.key === 'Enter') {
                onSearch(query)
              }
            }}
            onBlur={() => {
              if (query) {
                onSearch(query)
              } else {
                setIsEditing(false)
              }
            }}
            onClear={() => {
              setQuery('')
              onSearch('')
            }}
          />
        ) : (
          <Button isIconOnly radius="full" onClick={() => setIsEditing(true)} aria-label="Open Search">
            üîç
          </Button>
        )}
        {user && (
          <UserInfo
            user={user}
            onClick={() => {
              setIsEditing(true)
              setQuery(user.email)
              onSearch?.(user.email)
            }}
          />
        )}
      </div>
    </div>
  )
}

function formatRange(locale: string | undefined, start: string | undefined, end: string | undefined): string {
  const startDate = start ? new Date(start) : new Date()
  const endDate = end ? subDays(new Date(end), 1) : new Date()

  const format = new Intl.DateTimeFormat(locale, {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
  })
  return format.formatRange(startDate, endDate)
}

function getToday(locale: string | undefined) {
  const format = new Intl.DateTimeFormat(locale, {
    month: 'short',
    day: 'numeric',
  })
  return format.format(new Date())
}
