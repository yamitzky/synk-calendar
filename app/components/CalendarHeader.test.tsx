import { fireEvent, render, screen } from '@testing-library/react'
import { describe, expect, it, vi } from 'vitest'
import { CalendarHeader } from './CalendarHeader'

describe('CalendarHeader', () => {
  it('renders correctly', () => {
    const result = render(<CalendarHeader todayLabel="Today" />)
    expect(result).toMatchSnapshot()
  })

  it('calls onToday when today button is clicked', () => {
    const onToday = vi.fn()
    render(<CalendarHeader onToday={onToday} />)
    fireEvent.click(screen.getByLabelText('Today'))
    expect(onToday).toHaveBeenCalledTimes(1)
  })

  it('calls onNext and onPrev when navigation buttons are clicked', () => {
    const onNext = vi.fn()
    const onPrev = vi.fn()
    render(<CalendarHeader onNext={onNext} onPrev={onPrev} />)
    fireEvent.click(screen.getByLabelText('Next'))
    expect(onNext).toHaveBeenCalledTimes(1)
    fireEvent.click(screen.getByLabelText('Previous'))
    expect(onPrev).toHaveBeenCalledTimes(1)
  })

  it('displays custom todayLabel when provided', () => {
    const customLabel = 'navigate to today'
    render(<CalendarHeader todayLabel={customLabel} />)
    expect(screen.getByText(customLabel)).toBeInTheDocument()
  })

  it('calls onChangeView when view type is changed', () => {
    const onChangeView = vi.fn()
    render(<CalendarHeader onChangeView={onChangeView} initialView="timeGridDay" />)
    fireEvent.click(screen.getByLabelText('view type'))
    fireEvent.click(screen.getByTestId('timeGridWeek'))
    expect(onChangeView).toHaveBeenCalledWith('timeGridWeek')
  })

  it('displays correct date range', () => {
    render(<CalendarHeader start="2023-05-01" end="2023-06-01" />)
    expect(screen.getByRole('heading', { level: 1 })).toHaveTextContent('2023/05/01～2023/05/31')
  })

  it('displays correct date range when start and end are the same', () => {
    render(<CalendarHeader start="2023-05-01" end="2023-05-02" />)
    expect(screen.getByRole('heading', { level: 1 })).toHaveTextContent('2023年5月1日')
  })
})
