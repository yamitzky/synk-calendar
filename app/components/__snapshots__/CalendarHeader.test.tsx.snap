// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CalendarHeader > renders correctly 1`] = `
<div>
  <div
    class="grid gap-2 grid-cols-[repeat(2,max-content)] sm:grid-cols-[repeat(3,max-content)] justify-between items-center"
  >
    <div
      class="inline-flex items-center justify-center h-auto"
      role="group"
    >
      <button
        aria-label="Today"
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 min-w-20 h-10 text-small gap-2 [&:not(:first-child):not(:last-child)]:rounded-none [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-default text-default-foreground rounded-none first:rounded-s-medium last:rounded-e-medium data-[hover=true]:opacity-hover px-4"
        type="button"
      >
        Today
      </button>
      <button
        aria-label="Previous"
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-small gap-2 [&:not(:first-child):not(:last-child)]:rounded-none px-0 !gap-0 transition-transform-colors-opacity motion-reduce:transition-none bg-default text-default-foreground rounded-none first:rounded-s-medium last:rounded-e-medium min-w-10 w-10 h-10 data-[hover=true]:opacity-hover"
        type="button"
      >
        ◀
      </button>
      <button
        aria-label="Next"
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-small gap-2 [&:not(:first-child):not(:last-child)]:rounded-none px-0 !gap-0 transition-transform-colors-opacity motion-reduce:transition-none bg-default text-default-foreground rounded-none first:rounded-s-medium last:rounded-e-medium min-w-10 w-10 h-10 data-[hover=true]:opacity-hover"
        type="button"
      >
        ▶
      </button>
    </div>
    <h1
      class="text-lg text-center col-span-2 sm:col-span-1"
    >
      2023年5月1日
    </h1>
    <div
      class="w-40 row-start-1 col-start-2 sm:col-start-3"
    >
      <div
        class="group flex flex-col w-full transition-background motion-reduce:transition-none !duration-150 group relative justify-end data-[has-label=true]:mt-[calc(theme(fontSize.small)_+_10px)]"
        data-slot="base"
      >
        <div
          aria-hidden="true"
          data-a11y-ignore="aria-hidden-focus"
          data-testid="hidden-select-container"
          style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
        >
          <input
            style="font-size: 16px;"
            tabindex="-1"
            type="text"
            value=""
          />
          <label>
            <select
              size="4"
              tabindex="-1"
              title=""
            >
              <option />
              <option
                value="dayGridMonth"
              >
                月
              </option>
              <option
                value="timeGridWeek"
              >
                週
              </option>
              <option
                value="timeGridFourDay"
              >
                4日
              </option>
              <option
                value="timeGridDay"
              >
                日
              </option>
            </select>
          </label>
        </div>
        <div
          class="w-full flex flex-col"
          data-slot="mainWrapper"
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-label="view type"
            aria-labelledby="react-aria-:r7: react-aria-:r2:"
            class="relative px-3 gap-3 w-full inline-flex flex-row items-center shadow-sm tap-highlight-transparent group-data-[focus=true]:bg-default-100 h-10 min-h-10 rounded-medium bg-default-100 data-[hover=true]:bg-default-50 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2"
            data-slot="trigger"
            id="react-aria-:r2:"
            type="button"
          >
            <div
              class="inline-flex h-full w-[calc(100%_-_theme(spacing.6))] min-h-4 items-center gap-1.5 box-border"
              data-slot="innerWrapper"
            >
              <span
                class="text-foreground-500 font-normal w-full text-left rtl:text-right text-small truncate group-data-[has-value=true]:text-default-foreground"
                data-slot="value"
                id="react-aria-:r7:"
              />
            </div>
            <svg
              aria-hidden="true"
              class="absolute end-3 w-4 h-4 transition-transform duration-150 ease motion-reduce:transition-none data-[open=true]:rotate-180"
              data-slot="selectorIcon"
              fill="none"
              focusable="false"
              height="1em"
              role="presentation"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              width="1em"
            >
              <path
                d="m6 9 6 6 6-6"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
